include/rocksdb/sst_file_checksum.h

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+// Copyright (c) 2013 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+#pragma once
+
+#include <cassert>
+#include <map>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace rocksdb {
+
+// SstFileChecksum generates the checksum for each SST file when the file is
+// written to the file system. The checksum is stored in the Manifest
+class SstFileChecksum {
+ public:
+  virtual ~SstFileChecksum() {}
+  // Return the checksum of concat (A, data[0,n-1]) where init_checksum is the
+  // returned value of some string A. It is used to maintain the checksum of a
+  // stream of data
+  virtual uint32_t Extend(uint32_t init_checksum, const char* data,
+                          size_t n) = 0;
+
+  // Return the checksum value of data[0,n-1]
+  virtual uint32_t Value(const char* data, size_t n) = 0;
+
+  // Return a processed value of the checksum for store in somewhere
+  virtual uint32_t ProcessChecksum(const uint32_t checksum) = 0;
+
+  // Returns a name that identifies the current file checksum method.
+  virtual const char* Name() const = 0;
+};
+
+// The data structure that stores the mapping from file number to its checksum
+// It is used to store the file checksum from Manifest.
+struct FileChecksumList {
```

Comment:
Do we need to expose this class in public header in this PR? There is only one member function of `VersionSet` that uses this.

Reply:
@riversand963 Looks like it would be better to directly return an unordered_map from the function in VersionSet to avoid the introducing new class to user.

Reply:
Well, I think introducing a new class is fine, as long as we do not impose a certain data structure or implementation on the application. The public header should be the "interface", not implementation. Keep in mind that, it is much more difficult to remove something from public header once added.
Additionally, you can provide a default implementation of `FileChecksumList` whose implementation is also hidden. Applications can call `NewDefaultChecksumList()` to create it. The application can pass this to the upcoming API `DB::GetFileChecksums()`, which you will probably add in the next PR.
If the application is OK with default implementation of `FileChecksumList`, they can use it conveniently. Otherwise, they can create their own FileChecksumList by subclassing the base class defined in the public header.

Reply:
@riversand963 I will change accordingly. Thanks!